File: frontend/src/types/time-series.ts
/** Phase 18.2 Platform Time-Series Engine types. */

export type TimeSeriesInterval = "hour" | "day" | "week" | "month" | "quarter" | "on_demand";
export type CompareMode = "period" | "baseline" | "project" | "department" | "portfolio";

export type KpiObservation = {
  id: string;
  org_id: string;
  project_id: string | null;
  kpi_key: string;
  version: string;
  definition_version: string | null;
  calculator_key: string | null;
  calculator_version: string | null;
  observed_at: string;
  evaluated_at: string | null;
  numeric_value: number | string | null;
  text_value: string | null;
  normalized_value: number | string | null;
  confidence: number | string | null;
  value_type: string;
  status: string;
  department_key: string | null;
  agent_key: string | null;
  source_type: string;
  bucket_interval: string | null;
  bucket_start: string | null;
  bucket_end: string | null;
  evidence_refs: unknown[];
  lineage_refs: Record<string, unknown>;
  explainability: Record<string, unknown> | null;
  idempotency_fingerprint: string | null;
  supersedes_observation_id: string | null;
};

export type KpiTrendSummary = {
  kpi_key: string;
  latest: KpiObservation | null;
  previous: KpiObservation | null;
  absolute_change: number | string | null;
  percentage_change: number | string | null;
  raw_direction: "up" | "down" | "flat" | "unknown";
  semantic_favorability:
    | "improving"
    | "declining"
    | "stable"
    | "on_target"
    | "off_target"
    | "unknown";
  trend_direction_policy: string;
  observation_count: number;
  rolling_average: number | string | null;
  min_value: number | string | null;
  max_value: number | string | null;
  average_value: number | string | null;
  median_value: number | string | null;
};

export type KpiSeriesPoint = {
  bucket_start: string;
  bucket_end: string | null;
  numeric_value: number | string | null;
  text_value: string | null;
  observation_count: number;
  min_value: number | string | null;
  max_value: number | string | null;
  avg_value: number | string | null;
  median_value: number | string | null;
};

export type KpiSeries = {
  kpi_key: string;
  interval: TimeSeriesInterval;
  points: KpiSeriesPoint[];
  source: "observations" | "rollups";
};

export type KpiComparisonSeries = {
  label: string;
  scope_key: string;
  project_id: string | null;
  department_key: string | null;
  points: KpiSeriesPoint[];
  latest_value: number | string | null;
};

export type KpiCompare = {
  kpi_key: string;
  mode: CompareMode;
  interval: TimeSeriesInterval;
  baseline_label: string | null;
  series: KpiComparisonSeries[];
  absolute_deltas: Record<string, number | string | null>;
  percentage_deltas: Record<string, number | string | null>;
};

export type KpiForecastPoint = {
  forecast_at: string;
  value: number | string;
  lower_bound: number | string | null;
  upper_bound: number | string | null;
};

export type KpiForecast = {
  kpi_key: string;
  status: "ok" | "insufficient_data";
  method: string | null;
  model_version: string | null;
  horizon: number | null;
  training_window_start: string | null;
  training_window_end: string | null;
  sample_count: number;
  assumptions: string[];
  points: KpiForecastPoint[];
  message: string | null;
};

export type TimeSeriesDimensions = {
  kpi_keys: string[];
  agents: string[];
  departments: string[];
  intervals: string[];
  min_observed_at: string | null;
  max_observed_at: string | null;
};

export type RecommendationTimelineEvent = {
  id: string;
  org_id: string;
  project_id: string | null;
  domain: string;
  subject_table: string;
  subject_id: string;
  event_type: string;
  actor_user_id: string | null;
  source_agent: string | null;
  recommendation_type: string | null;
  severity: string | null;
  confidence: number | string | null;
  affected_kpi_keys: unknown[];
  status_snapshot: string | null;
  related_table: string | null;
  related_id: string | null;
  conversion_target: string | null;
  resolution_outcome: string | null;
  strategy_version: string | null;
  evidence_fingerprint: string | null;
  event_timestamp: string;
  payload: Record<string, unknown>;
};

export type RecommendationSubjectSummary = {
  domain: string;
  subject_table: string;
  subject_id: string;
  project_id: string | null;
  source_agent: string | null;
  recommendation_type: string | null;
  severity: string | null;
  status_snapshot: string | null;
  last_event_type: string;
  last_event_at: string;
  event_count: number;
};

export type TimeSeriesScopeParams = {
  org_id?: string;
  project_id?: string;
  department_key?: string;
  agent_key?: string;
  definition_version?: string;
  calculator_version?: string;
  date_from?: string;
  date_to?: string;
  interval?: TimeSeriesInterval;
  limit?: number;
  offset?: number;
  horizon?: number;
  rolling_window?: number;
  mode?: CompareMode;
  compare_project_id?: string;
};

File: backend/app/schemas/time_series.py
"""API schemas for the Phase 18.2 Platform Time-Series Engine."""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


IntervalLiteral = Literal["hour", "day", "week", "month", "quarter", "on_demand"]
CompareMode = Literal["period", "baseline", "project", "department", "portfolio"]


class TimeSeriesScopeFilters(BaseModel):
    org_id: UUID | None = None
    project_id: UUID | None = None
    client_user_id: UUID | None = None
    department_key: str | None = None
    agent_key: str | None = None
    definition_version: str | None = None
    calculator_version: str | None = None
    date_from: datetime | None = None
    date_to: datetime | None = None
    interval: IntervalLiteral | None = None


class KpiObservationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    org_id: UUID
    project_id: UUID | None = None
    kpi_key: str
    version: str
    definition_version: str | None = None
    calculator_key: str | None = None
    calculator_version: str | None = None
    observed_at: datetime
    evaluated_at: datetime | None = None
    numeric_value: Decimal | None = None
    text_value: str | None = None
    normalized_value: Decimal | None = None
    confidence: Decimal | None = None
    value_type: str = "numeric"
    status: str
    department_key: str | None = None
    agent_key: str | None = None
    source_type: str
    bucket_interval: str | None = None
    bucket_start: datetime | None = None
    bucket_end: datetime | None = None
    evidence_refs: list[Any] = Field(default_factory=list)
    lineage_refs: dict[str, Any] = Field(default_factory=dict)
    explainability: dict[str, Any] | None = None
    idempotency_fingerprint: str | None = None
    supersedes_observation_id: UUID | None = None


class KpiTrendSummaryRead(BaseModel):
    kpi_key: str
    latest: KpiObservationRead | None = None
    previous: KpiObservationRead | None = None
    absolute_change: Decimal | None = None
    percentage_change: Decimal | None = None
    raw_direction: Literal["up", "down", "flat", "unknown"] = "unknown"
    semantic_favorability: Literal[
        "improving", "declining", "stable", "on_target", "off_target", "unknown"
    ] = "unknown"
    trend_direction_policy: str
    observation_count: int = 0
    rolling_average: Decimal | None = None
    min_value: Decimal | None = None
    max_value: Decimal | None = None
    average_value: Decimal | None = None
    median_value: Decimal | None = None


class KpiSeriesPointRead(BaseModel):
    bucket_start: datetime
    bucket_end: datetime | None = None
    numeric_value: Decimal | None = None
    text_value: str | None = None
    observation_count: int = 1
    min_value: Decimal | None = None
    max_value: Decimal | None = None
    avg_value: Decimal | None = None
    median_value: Decimal | None = None


class KpiSeriesRead(BaseModel):
    kpi_key: str
    interval: IntervalLiteral
    points: list[KpiSeriesPointRead] = Field(default_factory=list)
    source: Literal["observations", "rollups"] = "observations"


class KpiComparisonSeriesRead(BaseModel):
    label: str
    scope_key: str
    project_id: UUID | None = None
    department_key: str | None = None
    points: list[KpiSeriesPointRead] = Field(default_factory=list)
    latest_value: Decimal | None = None


class KpiCompareRead(BaseModel):
    kpi_key: str
    mode: CompareMode
    interval: IntervalLiteral
    baseline_label: str | None = None
    series: list[KpiComparisonSeriesRead] = Field(default_factory=list)
    absolute_deltas: dict[str, Decimal | None] = Field(default_factory=dict)
    percentage_deltas: dict[str, Decimal | None] = Field(default_factory=dict)


class KpiForecastPointRead(BaseModel):
    forecast_at: datetime
    value: Decimal
    lower_bound: Decimal | None = None
    upper_bound: Decimal | None = None


class KpiForecastRead(BaseModel):
    kpi_key: str
    status: Literal["ok", "insufficient_data"]
    method: str | None = None
    model_version: str | None = None
    horizon: int | None = None
    training_window_start: datetime | None = None
    training_window_end: datetime | None = None
    sample_count: int = 0
    assumptions: list[str] = Field(default_factory=list)
    points: list[KpiForecastPointRead] = Field(default_factory=list)
    message: str | None = None


class TimeSeriesDimensionsRead(BaseModel):
    kpi_keys: list[str] = Field(default_factory=list)
    agents: list[str] = Field(default_factory=list)
    departments: list[str] = Field(default_factory=list)
    intervals: list[str] = Field(default_factory=list)
    min_observed_at: datetime | None = None
    max_observed_at: datetime | None = None


class RecommendationTimelineEventRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    org_id: UUID
    project_id: UUID | None = None
    domain: str
    subject_table: str
    subject_id: UUID
    event_type: str
    actor_user_id: UUID | None = None
    source_agent: str | None = None
    recommendation_type: str | None = None
    severity: str | None = None
    confidence: Decimal | None = None
    affected_kpi_keys: list[Any] = Field(default_factory=list)
    status_snapshot: str | None = None
    related_table: str | None = None
    related_id: UUID | None = None
    conversion_target: str | None = None
    resolution_outcome: str | None = None
    strategy_version: str | None = None
    evidence_fingerprint: str | None = None
    event_timestamp: datetime
    payload: dict[str, Any] = Field(default_factory=dict)


class RecommendationSubjectSummaryRead(BaseModel):
    domain: str
    subject_table: str
    subject_id: UUID
    project_id: UUID | None = None
    source_agent: str | None = None
    recommendation_type: str | None = None
    severity: str | None = None
    status_snapshot: str | None = None
    last_event_type: str
    last_event_at: datetime
    event_count: int = 1
